load analytics only when changed using a checksum

escherize opened the pull request:
resolves #38270

- over audit deserialization files

Extracts logic that decides to laod into `should-load-audit?` which checks the resource dir exists, `load-analytics-content` is turned on, and that the newly calculated checksum is different from the previously calculated checksum (calculated on startup each time).

<img width="398" alt="Screenshot 2024-01-30 at 2 46 24 PM" src="https://github.com/metabase/metabase/assets/343288/8260d2aa-5de3-4a61-a446-5f1db8336f53">

escherize committed:
load analytics only when changed using a checksum

- over audit deserialization files

noahmoss commented on enterprise/backend/src/metabase_enterprise/audit_db.clj:
```diff
@@ -186,24 +186,53 @@
-        (log/info (str "Loading Analytics Content Complete (" (count (:seen report)) ") entities loaded."))))
-    (when-let [audit-db (t2/select-one :model/Database :is_audit true)]
-      (adjust-audit-db-to-host! audit-db))))
+  (let [current-checksum (analytics-checksum)
+        last-checksum (last-analytics-checksum)]
```
maybe move these into `should-load-audit?`?

escherize replied:
I pulled this out to make it easier to test o:)

noahmoss commented on enterprise/backend/test/metabase_enterprise/audit_db_test.clj:
```diff
@@ -129,3 +129,16 @@
+    (is (= (#'audit-db/should-load-audit? #'audit-db/analytics-dir-resource true 1 3) true)))
+  (testing "checksums are the same => do not load"
+    (is (= (#'audit-db/should-load-audit? #'audit-db/analytics-dir-resource true 3 3) false)))
+  (testing
+      "resource path does not exist + load-analytics-content is false + checksums match  => do not load"
```
nit: whitespace

escherize replied:
This testing string was kinda long, so I broke it out

noahmoss commented on enterprise/backend/src/metabase_enterprise/audit_db.clj:
```diff
@@ -186,24 +186,53 @@
   :doc        false)
 
+(defsetting last-analytics-checksum
+  "Whether or not we should load Metabase analytics content on startup. Defaults to true, but can be disabled via environment variable."
```
nit: docstring out of date

escherize replied:
fixed

johnswanson commented on enterprise/backend/test/metabase_enterprise/audit_db_test.clj:
```diff
@@ -129,3 +129,16 @@
+    (is (= (#'audit-db/should-load-audit? #'audit-db/analytics-dir-resource true 3 3) false)))
+  (testing
+      "resource path does not exist + load-analytics-content is false + checksums match  => do not load"
+    (is (= (#'audit-db/should-load-audit? #'audit-db/analytics-dir-resource false 1 3) false))))
```
Should this say "checksums do not match", or be `3 3` instead of `1 3`?

escherize replied:
yep!

escherize committed:
update docstrings

escherize committed:
fix testing wording

escherize committed:
fix nil analytics-resource-dir

escherize committed:
in tests, when you delete the audit db, set the checksum back to 0

escherize committed:
add a comment

johnswanson committed:
try preventing loading analytics content in e2e

e2e tests are running extremely slowly with this change, and timing out.
As part of investigating what's going on, I'm going to disable analytics
loading in the e2e tests to see if that speeds them back up.

johnswanson committed:
parallelize calculation of analytics checksum

On my machine this was a ~3x speedup, from ~30ms to ~10ms.

johnswanson committed:
Add the ability to skip checksums

... while still loading analytics data.
